File: classifications/indoor.go
package classifications

import (
	"fmt"
	"math"
	"sync"

	"github.com/jatkinson1000/archeryutils/handicaps"
	"github.com/jatkinson1000/archeryutils/rounds"
)

type indoorGroupData struct {
	classes     []string
	classesLong []string
	classHC     []float64
}

var (
	indoorOnce sync.Once
	indoorDict map[string]indoorGroupData
)

var (
	indoorRoundsOnce sync.Once
	indoorRoundsMap  map[string]*rounds.Round
)

func allIndoorRounds() map[string]*rounds.Round {
	indoorRoundsOnce.Do(func() {
		indoorRoundsMap = make(map[string]*rounds.Round)
		for k, v := range rounds.AGBIndoor() {
			indoorRoundsMap[k] = v
		}
		for k, v := range rounds.WAIndoor() {
			indoorRoundsMap[k] = v
		}
	})
	return indoorRoundsMap
}

func buildIndoorDict() map[string]indoorGroupData {
	indoorOnce.Do(func() {
		agbAgeData := loadAgeData()
		agbBowstyleData := loadBowstyleData()
		classInfo := loadClassesIn()
		classes := classInfo.Classes
		classesLong := classInfo.ClassesLong
		n := len(classes)

		indoorDict = make(map[string]indoorGroupData)

		for _, bowstyle := range IndoorBowstyles {
			for _, gender := range AllGenders {
				for _, age := range AllAges {
					groupname := GroupName(bowstyle, gender, age)
					ageName := age.String()
					ad := agbAgeData[ageName]
					bd := agbBowstyleData[bowstyle.String()]

					ageCat := ad.Step
					delta := GetAgeGenderStep(gender, ageCat, bd.AgeStepIn, bd.GenderStepIn)
					hcs := make([]float64, n)
					for i := range n {
						hcs[i] = bd.DatumIn + delta + float64(i-1)*bd.ClassStepIn
					}

					indoorDict[groupname] = indoorGroupData{
						classes:     classes,
						classesLong: classesLong,
						classHC:     hcs,
					}
				}
			}
		}
	})
	return indoorDict
}

func CoaxIndoorGroup(bowstyle Bowstyle, gender Gender, age Age) Category {
	coaxedBowstyle := bowstyle
	switch bowstyle {
	case Flatbow, Traditional:
		coaxedBowstyle = Barebow
	case CompoundLimited, CompoundBarebow:
		coaxedBowstyle = Compound
	}
	return Category{Bowstyle: coaxedBowstyle, Gender: gender, AgeGroup: age}
}

func getIndoorGroupData(bowstyle Bowstyle, gender Gender, age Age) (indoorGroupData, error) {
	if !isIndoorBowstyle(bowstyle) {
		return indoorGroupData{}, fmt.Errorf(
			"%s is not a recognised bowstyle for indoor classifications", bowstyle)
	}
	if !isValidGender(gender) {
		return indoorGroupData{}, fmt.Errorf(
			"%s is not a recognised gender for indoor classifications", gender)
	}
	if !isValidAge(age) {
		return indoorGroupData{}, fmt.Errorf(
			"%s is not a recognised age group for indoor classifications", age)
	}
	dict := buildIndoorDict()
	groupname := GroupName(bowstyle, gender, age)
	if gd, ok := dict[groupname]; ok {
		return gd, nil
	}
	return indoorGroupData{}, fmt.Errorf("no indoor data for group %q", groupname)
}

func isIndoorBowstyle(b Bowstyle) bool {
	for _, ob := range IndoorBowstyles {
		if b == ob {
			return true
		}
	}
	return false
}

// CalculateIndoorClassification returns the classification string for a score.
func CalculateIndoorClassification(
	score float64,
	archeryRound *rounds.Round,
	bowstyle Bowstyle,
	gender Gender,
	age Age,
	strictRounds bool,
) (string, error) {
	if score < 0 || score > archeryRound.MaxScore() {
		return "", fmt.Errorf(
			"Invalid score of %.0f for a %s. Should be in range 0-%.0f.",
			score, archeryRound.Name, archeryRound.MaxScore(),
		)
	}

	scores, err := IndoorClassificationScores(archeryRound, bowstyle, gender, age, strictRounds)
	if err != nil {
		return "", err
	}

	gd, err := getIndoorGroupData(bowstyle, gender, age)
	if err != nil {
		return "", err
	}

	for i, classname := range gd.classes {
		classscore := float64(scores[i])
		if classscore < 0 || classscore > score {
			continue
		}
		return classname, nil
	}
	return "UC", nil
}

// IndoorClassificationScores returns score thresholds for each class (descending).
// -9999 indicates the class is not achievable on this round.
func IndoorClassificationScores(
	archeryRound *rounds.Round,
	bowstyle Bowstyle,
	gender Gender,
	age Age,
	strictRounds bool,
) ([]int, error) {
	gd, err := getIndoorGroupData(bowstyle, gender, age)
	if err != nil {
		return nil, err
	}

	allRounds := allIndoorRounds()
	roundname := archeryRound.Codename

	if strictRounds {
		if _, ok := allRounds[roundname]; !ok {
			return nil, fmt.Errorf(
				"This round is not recognised for indoor classification. "+
					"Please select an appropriate option. (codename=%q)", roundname)
		}
		if bowstyle == Compound {
			roundname = CompoundCodename(roundname)
		}
		archeryRound = allRounds[StripSpots(roundname)]
	}

	hcScheme := handicaps.MustScheme("AGB")
	n := len(gd.classes)
	classScores := make([]int, n)
	for i := range n {
		s := handicaps.ScoreForRound(hcScheme, gd.classHC[i], archeryRound, 0, true)
		classScores[i] = int(s)
	}

	// Gap/max-score handling: check score at floor(hc)+1
	for i, hc := range gd.classHC {
		nextScore := int(handicaps.ScoreForRound(hcScheme, math.Floor(hc)+1, archeryRound, 0, true))
		if nextScore == classScores[i] {
			if classScores[i] == int(archeryRound.MaxScore()) {
				classScores[i] = -9999
			} else {
				classScores[i]++
			}
		}
	}

	return classScores, nil
}

File: ts/src/classifications/indoor.ts
import { newScheme, scoreForRound } from '../handicaps.js';
import { agbIndoorRounds, waIndoorRounds, type Round, roundMaxScore } from '../rounds.js';
import {
  type Gender, type Age, type Bowstyle, type Category,
  AllGenders, AllAges, IndoorBowstyles,
  Gender as G, Bowstyle as B,
} from './data.js';
import { loadAgeData, loadBowstyleData, loadClassesIn } from './jsonio.js';
import { groupName, getAgeGenderStep, stripSpots, compoundCodename, fixRepeatedScores } from './utils.js';

interface IndoorGroupData {
  classes: string[];
  classesLong: string[];
  classHC: number[];
}

let indoorDict: Map<string, IndoorGroupData> | null = null;

function allIndoorRounds(): Map<string, Round> {
  const m = new Map<string, Round>();
  for (const [k, v] of agbIndoorRounds()) m.set(k, v);
  for (const [k, v] of waIndoorRounds()) m.set(k, v);
  return m;
}

function buildIndoorDict(): Map<string, IndoorGroupData> {
  if (indoorDict !== null) return indoorDict;
  const agbAgeData = loadAgeData();
  const agbBowstyleData = loadBowstyleData();
  const classInfo = loadClassesIn();
  const { classes, classes_long: classesLong } = classInfo;
  const n = classes.length;

  indoorDict = new Map();
  for (const bowstyle of IndoorBowstyles) {
    for (const gender of AllGenders) {
      for (const age of AllAges) {
        const gname = groupName(bowstyle, gender, age);
        const ad = agbAgeData[age];
        const bd = agbBowstyleData[bowstyle];
        const ageCat = ad.step;
        const delta = getAgeGenderStep(gender, ageCat, bd.ageStep_in, bd.genderStep_in);
        const hcs = Array.from({ length: n }, (_, i) => bd.datum_in + delta + (i - 1) * bd.classStep_in);
        indoorDict.set(gname, { classes, classesLong, classHC: hcs });
      }
    }
  }
  return indoorDict;
}

function isIndoorBowstyle(b: Bowstyle): boolean { return IndoorBowstyles.includes(b); }
function isValidGender(g: Gender): boolean { return g === G.Male || g === G.Female || g === G.Open; }
function isValidAge(a: Age): boolean { return AllAges.includes(a); }

function getIndoorGroupData(bowstyle: Bowstyle, gender: Gender, age: Age): IndoorGroupData {
  if (!isIndoorBowstyle(bowstyle))
    throw new Error(`${bowstyle} is not a recognised bowstyle for indoor classifications`);
  if (!isValidGender(gender))
    throw new Error(`${gender} is not a recognised gender for indoor classifications`);
  if (!isValidAge(age))
    throw new Error(`${age} is not a recognised age group for indoor classifications`);
  const gd = buildIndoorDict().get(groupName(bowstyle, gender, age));
  if (!gd) throw new Error(`No indoor data for group "${groupName(bowstyle, gender, age)}"`);
  return gd;
}

export function coaxIndoorGroup(bowstyle: Bowstyle, gender: Gender, age: Age): Category {
  let coaxed = bowstyle;
  if (bowstyle === B.Flatbow || bowstyle === B.Traditional) coaxed = B.Barebow;
  else if (bowstyle === B.CompoundLimited || bowstyle === B.CompoundBarebow) coaxed = B.Compound;
  return { bowstyle: coaxed, gender, ageGroup: age };
}

export function indoorClassificationScores(
  archeryRound: Round,
  bowstyle: Bowstyle,
  gender: Gender,
  age: Age,
  strictRounds: boolean,
): number[] {
  const gd = getIndoorGroupData(bowstyle, gender, age);
  const allRounds = allIndoorRounds();
  let round = archeryRound;
  let roundname = archeryRound.codename;

  if (strictRounds) {
    if (!allRounds.has(roundname)) {
      throw new Error(
        `This round is not recognised for indoor classification. Please select an appropriate option. (codename="${roundname}")`
      );
    }
    if (bowstyle === B.Compound) roundname = compoundCodename(roundname);
    const stripped = allRounds.get(stripSpots(roundname));
    if (stripped) round = stripped;
  }

  const hcScheme = newScheme('AGB');
  const maxScore = roundMaxScore(round);
  const classScores = gd.classHC.map(hc => Math.trunc(scoreForRound(hcScheme, hc, round, 0, true)));

  for (let i = 0; i < gd.classHC.length; i++) {
    const nextScore = Math.trunc(scoreForRound(hcScheme, Math.floor(gd.classHC[i]) + 1, round, 0, true));
    if (nextScore === classScores[i]) {
      classScores[i] = classScores[i] === Math.trunc(maxScore) ? -9999 : classScores[i] + 1;
    }
  }
  fixRepeatedScores(classScores, maxScore);

  return classScores;
}

export function calculateIndoorClassification(
  score: number,
  archeryRound: Round,
  bowstyle: Bowstyle,
  gender: Gender,
  age: Age,
  strictRounds: boolean,
): string {
  const maxScore = roundMaxScore(archeryRound);
  if (score < 0 || score > maxScore) {
    throw new Error(
      `Invalid score of ${Math.trunc(score)} for a ${archeryRound.name}. Should be in range 0-${Math.trunc(maxScore)}.`
    );
  }
  const scores = indoorClassificationScores(archeryRound, bowstyle, gender, age, strictRounds);
  const gd = getIndoorGroupData(bowstyle, gender, age);
  for (let i = 0; i < gd.classes.length; i++) {
    if (scores[i] < 0 || scores[i] > score) continue;
    return gd.classes[i];
  }
  return 'UC';
}
